### PR TITLE
[IMP] sale_promotion_programs: Apply coupon on first sale order only

### PR DESCRIPTION
**Pull request by ajaniszewska-dev**
moved here : https://github.com/camptocamp/sale-workflow/pull/7

Issue 155

### CODE REVIEW

File: sale_coupon_advanced_option/models/sale_coupon_program.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    )
+
+    def _check_promo_code(self, order, coupon_code):
+        if (
+            self.first_order_only
+            and len(
+                self.env["sale.order"].search(
+                    [
+                        ("partner_id", "=", order.partner_id.id),
+                        ("state", "!=", "cancel"),
```
Comment (santostelmo) [inline sale_coupon_advanced_option/models/sale_coupon_program.py]: Shouldn't we exclude quotation state also? `"state", "not int" ("cancel", "draft")`

Comment (ajaniszewska-dev) [reply sale_coupon_advanced_option/models/sale_coupon_program.py]: @santostelmo  good point but we discussed with @gurneyalex that we are on safe side when considering draft orders as well  - for example customer started to create order but didn't finish, order will be saved as a draft and can be restored and finished at any time. So we decided to exclude only cancelled orders.

Comment (santostelmo) [reply sale_coupon_advanced_option/models/sale_coupon_program.py]: I understand rule, if there is `draft` orders pending to be validated the customer can not benefit of the coupon.

**Review by santostelmo — APPROVED** [review]

File: sale_coupon_advanced_option/models/sale_coupon_program.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            return super()._check_promo_code(order, coupon_code)
+
+    def _filter_programs_by_sale_order_count(self, order):
+        if (
+            self.env["sale.order"].search_count(
+                [("partner_id", "=", order.partner_id.id), ("state", "!=", "cancel")]
+            )
+            <= 1
+        ):
+            return self.filtered(lambda program: program.first_order_only is True)
```
Comment (p-tombez) [inline sale_coupon_advanced_option/models/sale_coupon_program.py]: The value is already a boolean, you don't need to test it against another boolean value
```suggestion
            return self.filtered(lambda program: program.first_order_only)
```

**Review by i-vyshnevska — COMMENTED** [review]
comments left on old PR
the biggest worries about test case, probably not showing a real behavior 
https://github.com/OCA/sale-workflow/pull/1219

**Review by p-tombez — COMMENTED** [review]
Overall LG, just a small suggestion
One other thing : why this PR is on camptocamp/sale-workflow and not on OCA/sale-workflow ?

File: sale_coupon_advanced_option/models/sale_coupon_program.py (hunk 2 above)
Comment (ajaniszewska-dev) [reply sale_coupon_advanced_option/models/sale_coupon_program.py]: @p-tombez, code is fixed.
regarding why it is not OCA, that was @gurneyalex  decision as we all functionalities in one feature/branch

**Comment by ajaniszewska-dev** [comment]
**update:** moved code to sale_coupon_advanced as it should be in that addon

File: sale_coupon_advanced_option/tests/test_first_so_program_only.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+            "The discount should be a 50% discount",
+        )
+
+    def test_reused_code_promo_sale_order_program(self):
+
+        order1 = self.create_sale_order(self.partner3, {self.product_A: 5.0})
+        self.process_coupon(order1, "30_discount")
+
+        order2 = self.create_sale_order(self.partner3, {self.product_A: 5.0})
+        with self.assertRaises(UserError):
```
Comment (i-vyshnevska) [inline sale_coupon_advanced_option/tests/test_first_so_program_only.py]: here you need check which one error is raised

File: sale_coupon_advanced_option/tests/test_first_so_program_only.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        discounts = set(order2.order_line.mapped("name")) - {"Product A"}
+
+        self.assertEqual(
+            len(order2.order_line.ids), 1, "The order should contain 1 line"
+        )
+        self.assertEqual(
+            len(discounts),
+            0,
+            "the order should contain 1 Product A line and no discounts",
+        )
```
Comment (i-vyshnevska) [inline sale_coupon_advanced_option/tests/test_first_so_program_only.py]:
```suggestion
```
this test lines are redundant

**Review by i-vyshnevska — COMMENTED** [review]
generally looks good, needs rebasing on target branch and squashing

**Pull request closed without merging**